### PR TITLE
[fix](be) Truncate native ORC timestamps to microseconds

### PR DESCRIPTION
**Pull request by OIiveirra**
### What problem does this PR solve?

Issue Number: None

Related PR: #67904

Problem Summary:

Paimon ORC tables can store timestamp values with nanosecond precision, while Doris exposes timestamps at microsecond precision. The Paimon JNI reader truncates sub-microsecond digits, but the native ORC reader rounded them. For example, `2024-01-02 10:04:05.123456789` was exposed as `.123456` by JNI and `.123457` by the native reader, causing equality predicates to return inconsistent results.

The existing Paimon predicate conversion keeps `TIMESTAMP(7/8/9)` comparisons as residual predicates. This change makes the native ORC path use the same truncation semantics as Doris and the JNI reader for row decoding, stripe statistics, and ORC search argument bounds.

### Release note

Fix inconsistent timestamp values and predicate results between native ORC and JNI readers for sub-microsecond timestamps.

### Check List (For Author)

- Test
    - [x] Regression test
        - `./run-regression-test.sh --conf /tmp/yuze-regression-full-conf.groovy --run -d external_table_p0/paimon -s paimon_timestamp_types`
    - [x] Unit Test
        - `./run-be-ut.sh --run --filter=NewOrcReaderTest.*Timestamp*:NewOrcReaderTest.ReadTimestampNanosecondsTruncatesToMicroseconds`
    - [ ] Manual test
    - [ ] No need to test or manual test
- Behavior changed:
    - [x] Yes. Native ORC timestamps now truncate sub-microsecond digits consistently with Doris precision and the JNI reader.
    - [ ] No.
- Does this need documentation?
    - [x] No.
    - [ ] Yes.

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label

### CODE REVIEW

**Comment by hello-stephen** [comment]
Thank you for your contribution to Apache Doris.
Don't know what should be done next? See [How to process your PR](https://doris.apache.org/community/how-to-contribute/review-and-ci).

Please clearly describe your PR:
1. What problem was fixed (it's best to include specific error reporting information). How it was fixed.
2. Which behaviors were modified. What was the previous behavior, what is it now, why was it modified, and what possible impacts might there be.
3. What features were added. Why was this function added?
4. Which code was refactored and why was this part of the code refactored?
5. Which functions were optimized and what is the difference before and after the optimization?

**Review by Gabriel39 — DISMISSED** [review]

**Review by Gabriel39 — DISMISSED** [review]

**Comment by Gabriel39** [comment]
run buildall

**Comment by OIiveirra** [comment]
run external

**Comment by Gabriel39** [comment]
run buildall

**Comment by OIiveirra** [comment]
run buildall

**Comment by OIiveirra** [comment]
run buildall

**Comment by hello-stephen** [comment]
<details>
<summary>TPC-H: <b>Total hot run time: 27644 ms</b></summary>

```
machine: 'aliyun_ecs.c7a.8xlarge_32C64G'
scripts: https://github.com/apache/doris/tree/master/tools/tpch-tools
Tpch sf100 test result on commit e7dca64b45bc167479ca7f6454b52ed74e195b67, data reload: false

------ Round 1 ----------------------------------
============================================
q1	17737	3955	3904	3904
q2	2277	350	282	282
q3	10000	1416	818	818
q4	4689	481	345	345
q5	7479	826	554	554
q6	173	168	138	138
q7	735	797	588	588
q8	9322	1635	1552	1552
q9	5473	4182	4205	4182
q10	6749	1600	1364	1364
q11	432	270	243	243
q12	631	408	295	295
q13	18065	2628	2011	2011
q14	261	258	238	238
q15	q16	725	723	655	655
q17	1750	1080	952	952
q18	6624	5639	5589	5589
q19	1351	1272	936	936
q20	493	392	267	267
q21	5853	2496	2433	2433
q22	428	352	298	298
Total cold run time: 101247 ms
Total hot run time: 27644 ms

----- Round 2, with runtime_filter_mode=off -----
============================================
q1	4239	4219	4180	4180
q2	763	583	585	583
q3	4564	4987	4394	4394
q4	2266	2312	1481	1481
q5	4245	4127	4125	4125
q6	226	171	128	128
q7	1705	1614	1407	1407
q8	2468	2291	2156	2156
q9	7604	7693	7774	7693
q10	4339	4359	4053	4053
q11	588	403	382	382
q12	773	752	516	516
q13	2521	2804	2188	2188
q14	299	314	275	275
q15	q16	700	711	667	667
q17	7723	7293	7241	7241
q18	11957	11071	11834	11071
q19	1173	1071	1066	1066
q20	2223	2217	1971	1971
q21	5890	4806	4970	4806
q22	542	474	407	407
Total cold run time: 66808 ms
Total hot run time: 60790 ms
```
</details>

**Comment by hello-stephen** [comment]
<details>
<summary>TPC-DS: <b>Total hot run time: 153078 ms</b></summary>

```
machine: 'aliyun_ecs.c7a.8xlarge_32C64G'
scripts: https://github.com/apache/doris/tree/master/tools/tpcds-tools
TPC-DS sf100 test result on commit e7dca64b45bc167479ca7f6454b52ed74e195b67, data reload: false

query5	4316	592	446	446
query6	447	179	163	163
query7	4890	544	288	288
query8	324	186	182	182
query9	8787	3936	3917	3917
query10	472	305	280	280
query11	5869	2188	1984	1984
query12	161	102	101	101
query13	1251	550	410	410
query14	6512	4556	4222	4222
query14_1	4041	3985	4015	3985
query15	195	191	173	173
query16	999	474	429	429
query17	881	685	508	508
query18	2419	451	311	311
query19	191	179	134	134
query20	103	101	102	101
query21	219	129	114	114
query22	13140	13048	12816	12816
query23	15452	14469	14039	14039
query23_1	14137	14046	14120	14046
query24	7516	1680	1207	1207
query24_1	1220	1246	1200	1200
query25	513	405	335	335
query26	1266	301	158	158
query27	2699	541	336	336
query28	4613	1947	1970	1947
query29	1037	584	459	459
query30	308	232	202	202
query31	895	767	647	647
query32	156	98	96	96
query33	526	329	258	258
query34	1243	1166	649	649
query35	725	752	646	646
query36	794	793	705	705
query37	149	103	90	90
query38	1830	1786	1711	1711
query39	709	675	666	666
query39_1	649	639	654	639
query40	218	123	110	110
query41	73	69	69	69
query42	96	93	89	89
query43	335	351	297	297
query44	1349	685	689	685
query45	188	179	180	179
query46	1009	1137	700	700
query47	1504	1514	1377	1377
query48	395	428	306	306
query49	597	406	296	296
query50	945	348	266	266
query51	10753	10811	10606	10606
query52	89	89	75	75
query53	243	250	182	182
query54	257	215	197	197
query55	82	81	68	68
query56	252	248	233	233
query57	1425	1386	1272	1272
query58	250	205	217	205
query59	1972	2082	1820	1820
query60	323	227	214	214
query61	141	139	137	137
query62	392	313	270	270
query63	217	166	163	163
query64	2827	989	795	795
query65	4080	3969	3938	3938
query66	1818	425	308	308
query67	20399	19997	19628	19628
query68	3345	1539	954	954
query69	402	301	254	254
query70	920	879	862	862
query71	298	229	225	225
query72	2963	2694	2183	2183
query73	855	836	430	430
query74	4598	4452	4263	4263
query75	2300	2278	1927	1927
query76	2387	1084	685	685
query77	350	381	291	291
query78	9032	9065	8439	8439
query79	1203	1149	747	747
query80	511	470	360	360
query81	459	275	244	244
query82	342	163	121	121
query83	340	268	236	236
query84	313	141	107	107
query85	818	445	372	372
query86	325	235	231	231
query87	1982	1975	1836	1836
query88	3593	2661	2674	2661
query89	345	284	241	241
query90	1819	176	179	176
query91	161	149	131	131
query92	100	89	89	89
query93	1422	1366	832	832
query94	524	298	306	298
query95	663	461	328	328
query96	1056	753	325	325
query97	2457	2441	2316	2316
query98	203	184	183	183
query99	736	721	604	604
Total cold run time: 240932 ms
Total hot run time: 153078 ms
```
</details>

**Comment by hello-stephen** [comment]
<details>
<summary>ClickBench: <b>Total hot run time: 24.07 s</b></summary>

```
machine: 'aliyun_ecs.c7a.8xlarge_32C64G'
scripts: https://github.com/apache/doris/tree/master/tools/clickbench-tools
ClickBench test result on commit e7dca64b45bc167479ca7f6454b52ed74e195b67, data reload: false

query1	0.01	0.01	0.01
query2	0.10	0.05	0.04
query3	0.26	0.14	0.13
query4	1.61	0.15	0.14
query5	0.25	0.22	0.23
query6	1.16	0.95	0.93
query7	0.04	0.01	0.00
query8	0.05	0.04	0.03
query9	0.40	0.34	0.36
query10	0.57	0.58	0.55
query11	0.21	0.15	0.14
query12	0.19	0.15	0.15
query13	0.47	0.46	0.46
query14	0.94	0.95	0.93
query15	0.62	0.58	0.59
query16	0.31	0.34	0.35
query17	1.13	1.04	1.10
query18	0.20	0.19	0.20
query19	2.05	1.93	1.89
query20	0.02	0.01	0.02
query21	15.45	0.22	0.14
query22	4.78	0.05	0.05
query23	16.13	0.33	0.13
query24	3.02	0.44	0.32
query25	0.11	0.05	0.05
query26	0.72	0.20	0.16
query27	0.03	0.04	0.04
query28	3.54	0.77	0.35
query29	12.47	3.99	3.21
query30	0.28	0.15	0.16
query31	2.78	0.57	0.30
query32	3.22	0.60	0.50
query33	3.23	3.24	3.27
query34	15.60	3.92	3.30
query35	3.22	3.25	3.24
query36	0.56	0.43	0.42
query37	0.08	0.06	0.06
query38	0.05	0.04	0.04
query39	0.04	0.03	0.03
query40	0.17	0.15	0.14
query41	0.08	0.03	0.02
query42	0.05	0.03	0.03
query43	0.04	0.04	0.03
Total cold run time: 96.24 s
Total hot run time: 24.07 s
```
</details>

**Comment by OIiveirra** [comment]
run external

**Comment by hello-stephen** [comment]
# BE UT Coverage Report
Increment line coverage ` ` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/e7dca64b45bc167479ca7f6454b52ed74e195b67_e7dca64b45bc167479ca7f6454b52ed74e195b67/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/e7dca64b45bc167479ca7f6454b52ed74e195b67_e7dca64b45bc167479ca7f6454b52ed74e195b67/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 63.51% (29570/46562) |
| Line Coverage     | 48.24% (307966/638458) |
| Region Coverage   | 43.85% (248644/566977) |
| Branch Coverage   | 45.39% (115541/254546) |

**Comment by hello-stephen** [comment]
# BE Regression && UT Coverage Report
Increment line coverage `100% (0/0)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/68024_e7dca64b45bc167479ca7f6454b52ed74e195b67_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/68024_e7dca64b45bc167479ca7f6454b52ed74e195b67_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 76.27% (34390/45090) |
| Line Coverage     | 61.14% (385900/631186) |
| Region Coverage   | 57.53% (324852/564704) |
| Branch Coverage   | 58.30% (147932/253757) |

**Comment by hello-stephen** [comment]
# BE Regression && UT Coverage Report
Increment line coverage `100% (0/0)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/68024_e7dca64b45bc167479ca7f6454b52ed74e195b67_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/68024_e7dca64b45bc167479ca7f6454b52ed74e195b67_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 76.29% (34399/45090) |
| Line Coverage     | 61.17% (386115/631186) |
| Region Coverage   | 57.58% (325157/564704) |
| Branch Coverage   | 58.34% (148054/253757) |

**Comment by OIiveirra** [comment]
run buildall

**Comment by hello-stephen** [comment]
<details>
<summary>TPC-H: <b>Total hot run time: 27943 ms</b></summary>

```
machine: 'aliyun_ecs.c7a.8xlarge_32C64G'
scripts: https://github.com/apache/doris/tree/master/tools/tpch-tools
Tpch sf100 test result on commit 6b78de03c5ecfbb6cc0d72bf7407bb2bc0e258c4, data reload: false

------ Round 1 ----------------------------------
============================================
q1	17645	3842	3846	3842
q2	2318	354	310	310
q3	10007	1440	792	792
q4	4677	482	349	349
q5	7499	828	543	543
q6	187	173	138	138
q7	739	790	588	588
q8	9323	1733	1632	1632
q9	5681	4323	4252	4252
q10	6734	1678	1345	1345
q11	438	270	249	249
q12	635	416	290	290
q13	18077	2637	2000	2000
q14	261	255	235	235
q15	q16	741	731	672	672
q17	1846	1167	1022	1022
q18	6520	5644	5579	5579
q19	1258	1168	950	950
q20	478	412	258	258
q21	5826	3055	2593	2593
q22	448	365	304	304
Total cold run time: 101338 ms
Total hot run time: 27943 ms

----- Round 2, with runtime_filter_mode=off -----
============================================
q1	4549	4687	4482	4482
q2	809	591	601	591
q3	4815	5573	4667	4667
q4	2262	2392	1434	1434
q5	4637	4543	4555	4543
q6	222	173	125	125
q7	1810	1713	1473	1473
q8	2401	2165	2170	2165
q9	7484	7328	7236	7236
q10	4266	4197	3799	3799
q11	528	370	336	336
q12	709	711	499	499
q13	2317	2636	2002	2002
q14	262	278	246	246
q15	q16	668	684	658	658
q17	7322	6803	6642	6642
q18	11929	11070	11831	11070
q19	1111	1020	977	977
q20	2215	2211	1904	1904
q21	5456	4452	4589	4452
q22	539	455	397	397
Total cold run time: 66311 ms
Total hot run time: 59698 ms
```
</details>

**Comment by hello-stephen** [comment]
<details>
<summary>TPC-DS: <b>Total hot run time: 153534 ms</b></summary>

```
machine: 'aliyun_ecs.c7a.8xlarge_32C64G'
scripts: https://github.com/apache/doris/tree/master/tools/tpcds-tools
TPC-DS sf100 test result on commit 6b78de03c5ecfbb6cc0d72bf7407bb2bc0e258c4, data reload: false

query5	4334	584	464	464
query6	440	189	163	163
query7	4853	550	296	296
query8	328	176	169	169
query9	8761	4007	4029	4007
query10	454	313	274	274
query11	5824	2170	2004	2004
query12	149	98	93	93
query13	1252	593	419	419
query14	6473	4565	4275	4275
query14_1	4039	4012	4021	4012
query15	217	203	177	177
query16	983	450	424	424
query17	915	679	536	536
query18	2435	441	336	336
query19	202	188	138	138
query20	107	103	104	103
query21	222	142	117	117
query22	13073	12999	12837	12837
query23	15478	14643	13869	13869
query23_1	14116	14144	14134	14134
query24	7710	1669	1249	1249
query24_1	1264	1225	1227	1225
query25	519	413	339	339
query26	1261	320	156	156
query27	2740	544	330	330
query28	4573	1966	1963	1963
query29	1015	583	448	448
query30	330	231	201	201
query31	898	762	634	634
query32	141	90	93	90
query33	505	305	245	245
query34	1194	1130	647	647
query35	737	753	652	652
query36	802	825	736	736
query37	158	103	92	92
query38	1826	1763	1708	1708
query39	706	692	662	662
query39_1	657	656	665	656
query40	218	133	100	100
query41	67	64	63	63
query42	90	92	87	87
query43	339	346	308	308
query44	1387	692	697	692
query45	183	175	173	173
query46	1064	1152	678	678
query47	1494	1497	1386	1386
query48	386	404	293	293
query49	586	405	288	288
query50	1023	345	240	240
query51	10576	10474	10149	10149
query52	83	85	73	73
query53	232	260	176	176
query54	264	201	203	201
query55	77	72	68	68
query56	235	216	204	204
query57	1400	1404	1298	1298
query58	234	215	207	207
query59	1964	2076	1827	1827
query60	272	239	217	217
query61	146	144	148	144
query62	404	316	264	264
query63	213	177	175	175
query64	2813	984	828	828
query65	4035	3968	3918	3918
query66	1806	413	305	305
query67	20030	20367	20092	20092
query68	3164	1537	855	855
query69	406	289	264	264
query70	962	898	893	893
query71	295	224	229	224
query72	2981	2528	2176	2176
query73	813	752	452	452
query74	4615	4503	4317	4317
query75	2295	2290	1937	1937
query76	2312	1124	716	716
query77	359	382	288	288
query78	9101	8992	8520	8520
query79	1316	1175	721	721
query80	515	439	366	366
query81	457	283	240	240
query82	268	158	127	127
query83	264	261	241	241
query84	293	137	115	115
query85	789	458	379	379
query86	286	237	236	236
query87	1991	1963	1832	1832
query88	3598	2728	2710	2710
query89	324	290	251	251
query90	2146	175	176	175
query91	169	153	127	127
query92	102	81	91	81
query93	1370	1426	842	842
query94	530	336	289	289
query95	677	361	410	361
query96	1039	798	334	334
query97	2421	2447	2318	2318
query98	209	184	185	184
query99	745	703	611	611
Total cold run time: 240289 ms
Total hot run time: 153534 ms
```
</details>

**Comment by hello-stephen** [comment]
<details>
<summary>ClickBench: <b>Total hot run time: 24.03 s</b></summary>

```
machine: 'aliyun_ecs.c7a.8xlarge_32C64G'
scripts: https://github.com/apache/doris/tree/master/tools/clickbench-tools
ClickBench test result on commit 6b78de03c5ecfbb6cc0d72bf7407bb2bc0e258c4, data reload: false

query1	0.01	0.01	0.01
query2	0.09	0.05	0.05
query3	0.26	0.13	0.13
query4	1.61	0.14	0.14
query5	0.26	0.23	0.22
query6	1.16	1.00	0.98
query7	0.04	0.01	0.01
query8	0.05	0.04	0.03
query9	0.39	0.33	0.34
query10	0.54	0.54	0.55
query11	0.20	0.15	0.14
query12	0.18	0.15	0.15
query13	0.46	0.47	0.46
query14	0.96	0.93	0.94
query15	0.61	0.58	0.58
query16	0.31	0.34	0.32
query17	1.10	1.10	1.06
query18	0.22	0.20	0.20
query19	1.98	1.91	1.95
query20	0.02	0.01	0.01
query21	15.47	0.22	0.13
query22	4.82	0.06	0.05
query23	16.11	0.30	0.13
query24	3.00	0.45	0.31
query25	0.10	0.05	0.05
query26	0.75	0.20	0.15
query27	0.05	0.04	0.03
query28	3.50	0.71	0.35
query29	12.49	4.00	3.18
query30	0.29	0.15	0.16
query31	2.78	0.56	0.32
query32	3.23	0.59	0.48
query33	3.16	3.19	3.21
query34	15.61	3.97	3.30
query35	3.21	3.22	3.24
query36	0.56	0.44	0.43
query37	0.10	0.07	0.07
query38	0.05	0.04	0.04
query39	0.04	0.03	0.02
query40	0.17	0.15	0.14
query41	0.08	0.03	0.03
query42	0.04	0.03	0.03
query43	0.05	0.04	0.03
Total cold run time: 96.11 s
Total hot run time: 24.03 s
```
</details>

**Comment by hello-stephen** [comment]
# BE Regression && UT Coverage Report
Increment line coverage `100% (0/0)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/68024_6b78de03c5ecfbb6cc0d72bf7407bb2bc0e258c4_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/68024_6b78de03c5ecfbb6cc0d72bf7407bb2bc0e258c4_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 76.25% (34379/45090) |
| Line Coverage     | 61.16% (386026/631223) |
| Region Coverage   | 57.52% (324832/564729) |
| Branch Coverage   | 58.33% (148022/253787) |

**Review by Gabriel39 — DISMISSED** [review]
Reviewed decoding, timestamp statistics conversion, and SARG bounds at this commit. No blocking issues found. The relevant timestamp BE unit tests and Paimon timestamp regression passed on this head. As a non-blocking follow-up, consider adding SARG tests at exact millisecond boundaries and for fractional nanoseconds near the end of a second, where the upper bound crosses into the next second.

**Comment by OIiveirra** [comment]
run buildall

File: be/src/format_v2/orc/orc_search_argument.cpp [hunk 1]
```diff
@@ -1248,17 +1248,17 @@ ::orc::Literal ceil_timestamp_literal_to_millis(const ::orc::Literal& literal) {
...
-    // All raw ORC values in [lower, upper) round half-up to this Doris microsecond. Round these
+std::pair<::orc::Literal, ::orc::Literal> timestamp_truncation_bounds(
+        const ::orc::Literal& literal) {
+    constexpr int32_t NANOS_PER_MICROSECOND = 1000;
+    // All raw ORC values in [lower, upper) truncate to this Doris microsecond. Expand these
     // boundaries toward the side that enlarges the SARG match set because ORC statistics retain
     // only millisecond precision.
-    return {shift_timestamp_literal(literal, -HALF_MICROSECOND_NANOS),
-            shift_timestamp_literal(literal, HALF_MICROSECOND_NANOS)};
+    return {literal, shift_timestamp_literal(literal, NANOS_PER_MICROSECOND)};
```
[the remark this answers is not in the document]
Comment (OIiveirra) [reply be/src/format_v2/orc/orc_search_argument.cpp:1256]: Resolved in `8b1f7a577e2`. The obsolete exact-epoch exclusions were removed while retaining the negative-seconds fallback. Added an epoch predicate test that verifies a negative-only stripe is pruned; all 29 targeted timestamp BE tests pass.

File: regression-test/data/external_table_p0/paimon/paimon_timestamp_types.out [hunk 2]
```diff
@@ -5,6 +5,31 @@
 -- !c2 --
 1	2024-01-02T10:04:05.100	2024-01-02T10:04:05.120	2024-01-02T10:04:05.123	2024-01-02T10:04:05.123400	2024-01-02T10:04:05.123450	2024-01-02T10:04:05.123456	2024-01-02T10:04:05.123456	2024-01-02T10:04:05.123456	2024-01-02T10:04:05.123456	2024-01-02T10:04:05.100	2024-01-02T10:04:05.120	2024-01-02T10:04:05.123	2024-01-02T10:04:05.123400	2024-01-02T10:04:05.123450	2024-01-02T10:04:05.123456	2024-01-02T10:04:05.123456	2024-01-02T10:04:05.123456	2024-01-02T10:04:05.123456
 
+-- !ts9_eq_orc --
+1
+1
```
[the remark this answers is not in the document]
Comment (OIiveirra) [reply regression-test/data/external_table_p0/paimon/paimon_timestamp_types.out]: Resolved. The native ORC predicate path no longer duplicates the source row, and the generated snapshots now contain exactly one matching `id = 1` without using `DISTINCT`. The updated `paimon_timestamp_types` regression passes with both scanner modes.

File: regression-test/data/external_table_p0/paimon/paimon_timestamp_types.out [hunk 3]
```diff
@@ -53,6 +78,27 @@
 -- !c2 --
 1	2024-01-02T10:04:05.100	2024-01-02T10:04:05.120	2024-01-02T10:04:05.123	2024-01-02T10:04:05.123400	2024-01-02T10:04:05.123450	2024-01-02T10:04:05.123456	2024-01-02T10:04:05.123456	2024-01-02T10:04:05.123456	2024-01-02T10:04:05.123456	2024-01-02T10:04:05.100	2024-01-02T10:04:05.120	2024-01-02T10:04:05.123	2024-01-02T10:04:05.123400	2024-01-02T10:04:05.123450	2024-01-02T10:04:05.123456	2024-01-02T10:04:05.123456	2024-01-02T10:04:05.123456	2024-01-02T10:04:05.123456
 
+-- !ts9_eq_orc --
```
[the remark this answers is not in the document]
Comment (OIiveirra) [reply regression-test/data/external_table_p0/paimon/paimon_timestamp_types.out:77]: Resolved. Native ORC timestamp decoding now uses the same nanosecond-to-microsecond truncation contract as the JNI/Paimon path, and the generated snapshots produce the same values and row sets for both `force_jni_scanner` settings. The updated `paimon_timestamp_types` regression passes locally.

File: be/src/format_v2/orc/orc_reader.cpp [hunk 4]
```diff
@@ -502,22 +502,12 @@ std::optional<DateV2Value<DateTimeV2ValueType>> datetime_v2_from_orc_millis(
...
-    // Calendar bounds depend on the target timezone, so only reject arithmetic overflow here and
-    // let the converted value below decide whether the statistic is representable by Doris.
-    int64_t rounded_seconds;
-    if (__builtin_add_overflow(seconds, microseconds_with_carry / MICROS_PER_SECOND,
-                               &rounded_seconds)) {
-        return std::nullopt;
-    }
-    const auto microseconds = cast_set<uint64_t>(microseconds_with_carry % MICROS_PER_SECOND);
+    // the same truncation as row decoding so zone-map pruning observes identical values.
+    const auto microseconds =
```
[the remark this answers is not in the document]
Comment (OIiveirra) [reply be/src/format_v2/orc/orc_reader.cpp:507]: Resolved in `8b1f7a577e2`. Timestamp MIN/MAX pushdown now falls back to a row scan when ORC exposes the conservative missing-tail sentinels (`minimumNanos == 0` or `maximumNanos == 999999`). Added coverage for both `TIMESTAMP` and `TIMESTAMP_INSTANT`; all 29 targeted timestamp BE tests pass.

**Comment by hello-stephen** [comment]
<details>
<summary>TPC-H: <b>Total hot run time: 28021 ms</b></summary>

```
machine: 'aliyun_ecs.c7a.8xlarge_32C64G'
scripts: https://github.com/apache/doris/tree/master/tools/tpch-tools
Tpch sf100 test result on commit 8b1f7a577e2626863ee908a3e140a5e3b2ac457a, data reload: false

------ Round 1 ----------------------------------
============================================
q1	17641	3881	3847	3847
q2	2312	350	293	293
q3	10056	1423	796	796
q4	4691	478	347	347
q5	7520	834	555	555
q6	183	170	137	137
q7	735	781	587	587
q8	9314	1469	1533	1469
q9	5492	4238	4242	4238
q10	6758	1619	1359	1359
q11	434	293	250	250
q12	637	417	302	302
q13	18036	2615	2009	2009
q14	252	262	235	235
q15	q16	720	724	669	669
q17	1756	1183	1114	1114
q18	6527	5609	5539	5539
q19	1162	1206	1082	1082
q20	495	418	278	278
q21	5596	2989	2605	2605
q22	460	370	310	310
Total cold run time: 100777 ms
Total hot run time: 28021 ms

----- Round 2, with runtime_filter_mode=off -----
============================================
q1	4657	4845	4605	4605
q2	806	603	596	596
q3	4859	5255	4661	4661
q4	2275	2342	1512	1512
q5	4746	4330	4664	4330
q6	230	176	136	136
q7	1890	1706	1465	1465
q8	2396	2117	2008	2008
q9	7333	7220	7211	7211
q10	4296	4255	3866	3866
q11	522	396	358	358
q12	754	718	504	504
q13	2293	2640	1997	1997
q14	275	270	252	252
q15	q16	666	682	600	600
q17	7364	6736	6685	6685
q18	11884	10980	11796	10980
q19	1118	1044	1011	1011
q20	2213	2183	1911	1911
q21	5419	4527	4577	4527
q22	509	466	422	422
Total cold run time: 66505 ms
Total hot run time: 59637 ms
```
</details>

**Comment by hello-stephen** [comment]
<details>
<summary>TPC-DS: <b>Total hot run time: 154135 ms</b></summary>

```
machine: 'aliyun_ecs.c7a.8xlarge_32C64G'
scripts: https://github.com/apache/doris/tree/master/tools/tpcds-tools
TPC-DS sf100 test result on commit 8b1f7a577e2626863ee908a3e140a5e3b2ac457a, data reload: false

query5	4333	603	469	469
query6	512	191	175	175
query7	4816	541	297	297
query8	332	179	174	174
query9	8843	3997	3981	3981
query10	463	305	243	243
query11	5754	2243	2043	2043
query12	163	96	96	96
query13	1241	550	382	382
query14	6500	4558	4316	4316
query14_1	4028	3967	3957	3957
query15	202	192	176	176
query16	986	456	414	414
query17	887	685	544	544
query18	2432	474	333	333
query19	201	184	142	142
query20	114	104	107	104
query21	227	141	119	119
query22	13083	13022	12794	12794
query23	15691	14616	13907	13907
query23_1	14256	14159	14092	14092
query24	7422	1660	1205	1205
query24_1	1230	1220	1220	1220
query25	532	418	338	338
query26	1109	304	164	164
query27	2777	539	333	333
query28	4525	1967	1980	1967
query29	1025	589	456	456
query30	307	238	196	196
query31	895	763	631	631
query32	153	92	88	88
query33	502	312	230	230
query34	1164	1118	629	629
query35	719	771	648	648
query36	793	817	724	724
query37	172	102	93	93
query38	1849	1775	1717	1717
query39	706	681	657	657
query39_1	631	660	644	644
query40	223	122	96	96
query41	70	70	68	68
query42	97	98	94	94
query43	342	347	302	302
query44	1349	689	691	689
query45	191	186	172	172
query46	1041	1162	689	689
query47	1482	1490	1387	1387
query48	378	414	280	280
query49	601	440	304	304
query50	975	359	244	244
query51	10547	10491	10519	10491
query52	82	89	75	75
query53	239	266	180	180
query54	257	217	193	193
query55	83	77	73	73
query56	262	228	239	228
query57	1587	1376	1308	1308
query58	240	223	218	218
query59	2013	2065	1828	1828
query60	287	255	236	236
query61	169	165	172	165
query62	406	320	273	273
query63	227	182	176	176
query64	2713	1121	929	929
query65	4044	3978	3987	3978
query66	1800	440	322	322
query67	20249	20095	20019	20019
query68	3567	1507	959	959
query69	425	295	273	273
query70	1007	899	887	887
query71	310	256	224	224
query72	2989	2461	2189	2189
query73	802	809	414	414
query74	4639	4499	4310	4310
query75	2294	2283	1954	1954
query76	2424	1148	705	705
query77	357	405	310	310
query78	9106	9059	8543	8543
query79	1283	1227	721	721
query80	608	468	353	353
query81	475	284	245	245
query82	610	174	125	125
query83	357	277	240	240
query84	315	141	109	109
query85	813	470	380	380
query86	336	243	239	239
query87	2014	1966	1850	1850
query88	3610	2731	2683	2683
query89	350	288	244	244
query90	1910	180	179	179
query91	170	158	128	128
query92	99	93	88	88
query93	1430	1510	915	915
query94	529	340	299	299
query95	664	385	338	338
query96	1091	814	366	366
query97	2470	2447	2297	2297
query98	204	199	187	187
query99	738	716	612	612
Total cold run time: 241876 ms
Total hot run time: 154135 ms
```
</details>

**Comment by hello-stephen** [comment]
<details>
<summary>ClickBench: <b>Total hot run time: 24.08 s</b></summary>

```
machine: 'aliyun_ecs.c7a.8xlarge_32C64G'
scripts: https://github.com/apache/doris/tree/master/tools/clickbench-tools
ClickBench test result on commit 8b1f7a577e2626863ee908a3e140a5e3b2ac457a, data reload: false

query1	0.01	0.01	0.01
query2	0.09	0.05	0.05
query3	0.26	0.14	0.14
query4	1.61	0.14	0.14
query5	0.25	0.22	0.21
query6	1.16	0.90	0.92
query7	0.04	0.01	0.01
query8	0.06	0.04	0.04
query9	0.39	0.33	0.33
query10	0.56	0.55	0.58
query11	0.20	0.15	0.15
query12	0.18	0.15	0.15
query13	0.47	0.47	0.48
query14	0.96	0.95	0.94
query15	0.63	0.58	0.60
query16	0.30	0.31	0.33
query17	1.06	1.06	1.08
query18	0.22	0.21	0.20
query19	2.04	1.88	1.93
query20	0.02	0.02	0.01
query21	15.48	0.21	0.13
query22	4.87	0.05	0.05
query23	16.14	0.32	0.12
query24	2.94	0.43	0.30
query25	0.10	0.06	0.04
query26	0.74	0.21	0.17
query27	0.04	0.04	0.04
query28	3.44	0.82	0.37
query29	12.50	4.15	3.22
query30	0.30	0.17	0.16
query31	2.77	0.58	0.31
query32	3.23	0.59	0.50
query33	3.15	3.30	3.26
query34	15.47	3.97	3.28
query35	3.24	3.20	3.21
query36	0.55	0.46	0.42
query37	0.09	0.06	0.07
query38	0.05	0.05	0.04
query39	0.04	0.03	0.03
query40	0.17	0.16	0.15
query41	0.09	0.04	0.03
query42	0.04	0.04	0.03
query43	0.05	0.04	0.04
Total cold run time: 96 s
Total hot run time: 24.08 s
```
</details>

**Comment by hello-stephen** [comment]
# BE UT Coverage Report
Increment line coverage ` ` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/8b1f7a577e2626863ee908a3e140a5e3b2ac457a_8b1f7a577e2626863ee908a3e140a5e3b2ac457a/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/8b1f7a577e2626863ee908a3e140a5e3b2ac457a_8b1f7a577e2626863ee908a3e140a5e3b2ac457a/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 63.52% (29554/46529) |
| Line Coverage     | 48.24% (307693/637884) |
| Region Coverage   | 43.87% (248423/566330) |
| Branch Coverage   | 45.40% (115416/254223) |

**Comment by hello-stephen** [comment]
# BE Regression && UT Coverage Report
Increment line coverage `100% (0/0)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/68024_8b1f7a577e2626863ee908a3e140a5e3b2ac457a_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/68024_8b1f7a577e2626863ee908a3e140a5e3b2ac457a_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 76.31% (34379/45054) |
| Line Coverage     | 61.17% (385737/630604) |
| Region Coverage   | 57.56% (324685/564048) |
| Branch Coverage   | 58.34% (147864/253434) |

**Review by Gabriel39 — APPROVED** [review]
